lib/relooper/src/graph/caterpillar.rs

```diff
@@ -1,166 +1,118 @@
+    cfg_label: T,
     is_dynamic: bool,
     is_jumpdest: bool,
 }
 
 #[derive(PartialOrd, PartialEq, Clone, Copy, Hash, Eq, Ord)]
-pub enum CaterpillarLabel {
-    Original(usize),
-    Generated(usize, usize), // (unique_id, offset of associated jumpdest)
+pub enum CaterpillarLabel<T> {
+    Original(T),
+    Generated(T), // (unique_id, offset of associated jumpdest)
 }
 
-impl Display for CaterpillarLabel {
+impl<T: Display> Display for CaterpillarLabel<T> {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
         match &self {
-            CaterpillarLabel::Generated(id, offset) => write!(f, "{}_{}", id, offset),
+            CaterpillarLabel::Generated(offset) => write!(f, "g{}",offset),
             CaterpillarLabel::Original(id) => write!(f, "{}", id),
         }
     }
 }
 
-pub fn unfold_dyn_edges(cfg: Cfg<EvmLabel>) -> Cfg<CaterpillarLabel> {
-    let to_caterpillar_edge = |(label, edge): (&EvmLabel, &CfgEdge<EvmLabel>)| -> (CaterpillarLabel, CfgEdge<CaterpillarLabel>) {
-        match edge {
-            CfgEdge::Cond(cond, uncond) => {
-                (
-                    CaterpillarLabel::Original(label.cfg_label),
-                    CfgEdge::Cond(
-                        CaterpillarLabel::Original(cond.cfg_label),
-                        CaterpillarLabel::Original(uncond.cfg_label),
-                    )
-                )
-            }
-            CfgEdge::Uncond(uncond) => {
-                (
-                    CaterpillarLabel::Original(label.cfg_label),
-                    CfgEdge::Uncond(CaterpillarLabel::Original(uncond.cfg_label)),
-                )
-            }
-            CfgEdge::Terminal => {
-                (
-                    CaterpillarLabel::Original(label.cfg_label),
-                    CfgEdge::Terminal,
-                )
-            }
-        }
-    };
-    let mut outedges: HashMap<CaterpillarLabel, CfgEdge<CaterpillarLabel>> = cfg
-        .out_edges
-        .iter()
-        .filter(|(label, _)| !label.is_dynamic)
-        .map(to_caterpillar_edge)
-        .collect();
-    let jumpdests: Vec<usize> = cfg
-        .out_edges
-        .iter()
-        .filter_map(|(label, _)| {
-            if label.is_jumpdest {
-                Some(label.cfg_label)
-            } else {
-                None
+pub fn unfold_dyn_edges<T: CfgLabel>(cfg: Cfg<EvmLabel<T>>) -> Cfg<CaterpillarLabel<T>> {
+    let mut cat_cfg = cfg.map_label(|&label| CaterpillarLabel::Original(label.cfg_label));
+
+    let dyn_nodes: Vec<_> = cfg.nodes().into_iter().filter(|l| l.is_dynamic).collect();
+    let jumpdests: Vec<_> = cfg.nodes().into_iter().filter(|l| l.is_jumpdest).collect();
+
+    match &jumpdests[..] {
+        [] => cat_cfg,
+        [single] => {
+            let single_edge = CaterpillarLabel::Original(single.cfg_label);
+            for d in dyn_nodes {
+                cat_cfg.add_edge(CaterpillarLabel::Original(d.cfg_label), CfgEdge::Uncond(single_edge));
             }
```

Comment:
I think here you don't take consideration about existence of cases when node has conditional dynamic edge and unconditional static edge.